ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -587,6 +588,32 @@ releaseResources rr keys releaser = do
 withRegistry :: (IOLike m, HasCallStack) => (ResourceRegistry m -> m a) -> m a
 withRegistry = bracket unsafeNewRegistry closeRegistry
 
+-- | Create a new registry as part of a standard 'bracket'
+--
+-- Use this combinator to emphasize that the lifetime of the registry exactly
+-- spans the lifetime of the explicitly bracketed resources, no more and no
+-- less.
+--
+-- The only difference from 'bracketWithRegistry_' is that the explicitly
+-- bracketed resources do rely on the registry.
+--
+-- See documentation of 'ResourceRegistry' for a more general discussion.
+bracketWithRegistry ::
+     (IOLike m, HasCallStack)
+  => (ResourceRegistry m -> m a)
+     -- ^ Allocate the bracketed resources. They depend on the registry but are
+     -- not themselves registered, since they're simply 'bracket'ed.
+  -> (a -> m ())
+     -- ^ Release the bracketed resources.
+  -> (ResourceRegistry m -> a -> m r)
+  -> m r
+bracketWithRegistry newA closeA body =
+    -- It would be OK to fuse these two calls to 'bracket', but it's simpler and
+    -- less error-prone this way.
+    withRegistry $ \registry ->
```

Comment:
We can side-step the precondition that @kderme's noticed (nice catch!) by simply allocating the `a` in the registry. This combinator ensures it's the oldest thing in the registry, and so it will be deallocated last.

Reply:
Also rename to `brackWithPrivateRegistry` and remove the `registry` argument from the `body`: none of our uses cases need it yet, so let's simplify at least for now.

The fact that none of our use cases need it yet suggested that perhaps `newFollower` should simply be replaced with `withFollower`, but it turns out that the QSM tests need the `newFollower` variant, since QSM inherently uses a flat sequence of commands: we'd have no hook into QSM in which to call `withFollower` with a scope other than "the whole test case".

Reply:
Edsko and I discussed going farther and not passing the `registry` to the body. However, I just realized that the `a` type variable could be abused to pass the registry to the body.

So I'll still make no assumptions about orderings: we'll allocate the "bracketed resources" (via `newA`) in the registry.

Reply:
I also just realized that the `newA` argument might itself allocate something in the registry. This would be weird, but allowed by is allowed by the types so far. I'm about to push commits that use `m (ResourceRegistry m -> a)` instead, which prevents that possibility at the cost of a slightly more refactoring.

Reply:
We backed out those new commits. We're allowing `newA` to allocate things in the registry. (I don't think it'd even _need_ to do that, since it could just close them in the `closeA`. But that might require extensions to the `a` interface.)

Reply:
And because `newA` may allocate resources that'll be younger than the `a` bracketed here, we must track `a` in the registry; we can't close it separate from the registry because the registry has some older than `a` and some younger than `a` resources.